### ISSUE TITLE
Fix the `TODO` in `pauseDiamond` function

### ISSUE THREAD

**Issue by sujithsomraaj** [sujithsomraaj/lifi-emergency-pause-facet-v1.2.0#4]
**Context:** [EmergencyPauseFacet.sol#L84](https://github.com/lifinance/contracts/blob/77441a088e0789513db4e068f7ef6c5c0988ee42/src/Facets/EmergencyPauseFacet.sol#L84)

**Description:** The `pauseDiamond` function is used by the pauser wallet/diamond admin to pause the diamond contract by removing all the function selectors. To do so, the function queries all the facets of the diamond using the `_getAllFacetFunctionSelectorsToBeRemoved` internal functions. 

However, this function risks running out of gas if the size of the facets grows over time, and the same is added as a `TODO` in the function, which is not completed. 

```solidity
//TODO: add handling for cases with too many facets, and TX will run out of gas (>> pagination).
```

**Recommendation:** Consider adding a new function that allows the admin to pass an array of facets to pause, which can be used to pause the diamond without running out of block gas.

**LI.FI:** Acknowledged. 

**Researcher:** Acknowledged. It is unlikely there will be as many facets to run into the out-of-gas issue; if it happens, a new emergency pause facet has to be added. Adding a watcher to the deploy pipeline to make sure whenever you add a new facet, you can pause the entire diamond.

**Comment by 0xDEnYO** [comment]
Hey Sujith, 

I am hesitant with this finding. I had such a function in there before but I removed it for the following reasons: 
- I wanted the pause function to be as simple as possible (no parameters) 
- I wanted to be able to use the same calldata for pausing across all networks (which would not be possible if we had to pass in a list of facets as they differ on each network)
- I want to prevent that the contract can be partially paused (e.g. by adding an incomplete list of facets)

I know that it's an admin-only function but it just adds complexity to have to get the list of facets first and then pass it as parameter, so I tried to avoid it.

My solution to the potential out-of-gas issue was to add a second test file `EmergencyPauseFacet.fork.t.sol` which simulates the `EmergencyPauseFacet` to be added to our production diamond on mainnet (which has most facets) and then tries to pause it. 
That means that as soon as we reach a state where this function would indeed run out of gas, our unit test suite would fail and thereby raise our attention to this issue.

@maxklenk
How do you feel about this topic?
Do you prefer to add this second function immediately or would you be OK with the current approach and only add this function when it's really necessary?

**Comment by maxklenk** [comment]
I agree with your arguments why it makes sense to be able to pause the contract in one call.
If possible I would like to stick with it, but it is critical that this function works at all times.

The test case is a good idea to check if we are still in the limit of an executable transaction. I would like to know at which size we are facing problems. Can we calculate/test how many facets + functions we can have before we would hit a size limit? Can we have gas optimizations, by e.g. calling `LibDiamond.replaceFunctions` only once with all `functionSelectors` or storing all facets at once `s.facets = facets`?

**Comment by ezynda3** [comment]
Agreed. Would stick with the simpler interface and monitor as well as explore ways to gas optimize this call in the future.

**Comment by 0xDEnYO** [comment]
> The test case is a good idea to check if we are still in the limit of an executable transaction. I would like to know at which size we are facing problems. Can we calculate/test how many facets + functions we can have before we would hit a size limit? Can we have gas optimizations, by e.g. calling `LibDiamond.replaceFunctions` only once with all `functionSelectors` or storing all facets at once `s.facets = facets`?

I dont think that we can gas optimize this way.
We dont have a simple way to just retrieve all functionSelectors (without the association to the facetAddress) as they are stored in a mapping (with the key being the facetAddress). 

Also `s.facets = facets` wont work since copying memory to storage arrays like that is not supported in Solidity. 


I will write a test case but conclude this issue with our decision that the test case which monitors the executability (as well as regular simulation runs) is sufficient for our purpose.

**Comment by 0xDEnYO** [comment]
@maxklenk 
I've added a test case that creates dummy contracts with each 3 random function selectors and adds it to the diamond, then unpauses it. 
Ran it with 500 contracts and it passed. 
I dont really know how representative that is (i.e. if local foundry testnet has similar max gas limit than a mainnet).
For now I feel OK with relying on the ETH mainnet fork test case plus quarterly executed simulations in Tenderly. 
Ok for you?

**Comment by maxklenk** [comment]
Thanks, that is a good reference, I think in this case it should not be a problem and we can remove the TODO. Once it is live we should run a simulation with gas settings on each chains real rpc to see if the transaction would work.

**Comment by 0xDEnYO** [comment]
Todo removed an comment added to explain our reasoning
